app/src/dex/router/fill_route.rs

```diff
@@ -188,11 +210,20 @@ pub trait FillRoute: StateWrite + Sized {
             }
 
             if current_value.amount == 0u64.into() {
-                println!("zero current value.");
-                // Note: this can be hit during dust fills
-                // TODO(erwan): craft `test_dust_fill_zero_value` to prove this.
-                panic!("zero current value");
-                break 'filling;
+                // While the `input` to `fill_route` can be zero, the 'filling
+                // loop requires that `input.amount > 0`. The only other possible
+                // source of a `current_value.amount == 0` is from the saturating
+                // input calculation (lifting the limiting constraint).
+                // There are two reasons why no saturating can be zero:
+                // + delta_1_star = ceil(delta_1), so unless delta_1 == 0, we
+                // have delta_1_star > 0.
+                // + delta_1 = lambda_2 * accumulated_effective_price
+                // And lambda_2 != 0, since lambda_2 = r2 and r2 != 0 because
+                // otherwise it would not have been indexed for that pair's
+                // direction.
+                // + accumulated_effective_price != 0, because for each position
+                //   we have p, q != 0.
+                unreachable!("current_value is nonzero")
```

Comment:
@hdevalence you were right! it should be impossible to hit that branch, which is a relief. In the process of proving why i realized that a much more load bearing aspect of routing is in the end-to-end effective price calculation, so this is the next item on my TODO of double-checking assumptions. They are already marked as TODOs in the code.